tests/collections/llm/test_fault_nvrx.py

```diff
@@ -0,0 +1,126 @@
+    )
+    parser.add_argument(
+        '--experiment-dir', type=str, required=True, help="directory to write results and checkpoints to"
+    )
+    parser.add_argument(
+        '--data-path', type=str, default=None, help="Path to data file. If not specified, uses mock data."
+    )
+    parser.add_argument(
+        '--tokenizer-path',
+        type=str,
+        default=None,
+        help="Path to a sentencepiece tokenizer model file. If not specified, uses mock data.",
+    )
+    parser.add_argument('--index-mapping-dir', type=str, help="directory to write index mappings to")
+
+    return parser.parse_args()
+
+
+def main():
+    args = get_args()
+
+    exp_name = "L2_llama3_small_pretrain_fault_tolerance_test"
+    pretrain_recipe = llm.llama3_8b.pretrain_recipe(
+        dir=args.experiment_dir, name=exp_name, num_gpus_per_node=args.devices
+    )
+
+    pretrain_recipe.model = run.Config(llm.LlamaModel, small_llama_cfg(1024))
+
+    if args.data_path and args.tokenizer_path:
+        pretrain_recipe.data = train_data(
+            data_path=args.data_path,
+            tokenizer_path=args.tokenizer_path,
+            index_mapping_dir=args.index_mapping_dir,
+            seq_length=1024,
+        )
+
+    # Recipe Overrides
+    pretrain_recipe.trainer.max_steps = 20
+    pretrain_recipe.trainer.log_every_n_steps = 1
+    pretrain_recipe.log.ckpt.every_n_train_steps = None
+    pretrain_recipe.log.ckpt.train_time_interval = None
+    pretrain_recipe.trainer.val_check_interval = 30
+    pretrain_recipe.trainer.limit_val_batches = 2
+
+    executor: run.SlurmExecutor = run.LocalExecutor(ntasks_per_node=args.devices, launcher="ft")
+    # Add the fault tolerance plugin which enables restart after a crash
+    run_plugins: list[run.Plugin] = [FaultTolerancePlugin(num_in_process_restarts=1, num_job_retries_on_failure=0)]
+    pretrain_recipe.trainer.callbacks = [
+        run.Config(TimingCallback),
+        straggler_det_callback(straggler_report_time_interval=0.5),
+    ]
+
+    if args.crash_step:
+        pretrain_recipe.trainer.callbacks.append(run.Config(CrashCallback, crash_step=args.crash_step))
+
+    run.run(pretrain_recipe, plugins=run_plugins, executor=executor)
+
+    # Assumes that NeMo logs are written into "run.log"
+    # When a crash a simulated, error shows up on the terminal but it is not written to a file
+    # So the test appends run output to run.log in the experiment-dir
+    log_content = None
+    with open(os.path.join(args.experiment_dir, "run.log")) as f:
+        log_content = f.read()
+
+    if args.check_report:
+        assert "GPU relative performance" in log_content
+        assert "GPU individual performance" in log_content
+        assert "Straggler report processing time" in log_content
+    if args.crash_step:
+        assert f"Exception: Simulating a crash at step {args.crash_step}!" in log_content
+
```

Comment:
We should somehow confirm that we still finish the training despite the crash, right? Maybe the common.verify_ckpt_dir will be enough, it should ensure that there's a checkpoint for step 20.